src/util/constraints.rs

```diff
@@ -90,7 +236,87 @@ async fn validate_batch_with_constraint(
         .fail()?;
     }
 
-    Ok(())
+    // Return the processed batches
+    let final_batches = df.collect().await.context(DataFusionSnafu)?;
+    Ok(final_batches)
+}
+
+/// Add row numbers to batches for tracking insertion order in last write wins logic
+fn add_row_numbers_to_batches(batches: Vec<RecordBatch>) -> Result<Vec<RecordBatch>> {
+    let mut batches_with_row_nums = Vec::new();
+    let mut global_row_number = 0i64;
+
+    for batch in batches {
+        let num_rows = batch.num_rows();
+        let row_numbers: Vec<i64> =
+            (global_row_number..global_row_number + num_rows as i64).collect();
+        global_row_number += num_rows as i64;
+
+        // Create new schema with additional row_num column
+        let mut fields = batch.schema().fields().iter().cloned().collect::<Vec<_>>();
+        fields.push(Arc::new(Field::new("__row_num", DataType::Int64, false)));
```

Comment:
```suggestion
        fields.push(Arc::new(Field::new(BATCH_ROW_NUMBER_COLUMN_NAME, DataType::Int64, false)));
```

nit: magic column name